src/vos/vos_dtx.c

```diff
@@ -1049,11 +1049,9 @@ vos_dtx_alloc(struct umem_instance *umm, struct dtx_handle *dth)
 	if (rc == 0) {
 		dth->dth_ent = dae;
 		dth->dth_active = 1;
-	}
-
-out:
-	if (rc != 0)
+	} else {
 		dtx_evict_lid(cont, dae);
+	}
```

Comment:
I made this change in my punch propagation patch:
![image](https://user-images.githubusercontent.com/10464486/95758750-139a2a00-0c66-11eb-87bf-572955012248.png)
I wonder if it's the same issue you're running into?   The issue for me was on retry, if I aborted the first transaction, the dth_ent was still set...but it was a unit test so maybe it's unique to my situation.

Reply:
My issue is that when server out of space, when vos_dtx_alloc() calls vos_dtx_extend_act_table(), it will get "-DER_NOSPACE". At that time, it has allocated the DTX entry via lrua_allocx() but not initialized, then vos_dtx_alloc() will call dtx_evict_lid(), and then trigger "D_ASSERT(key != 0);" inside lrua_evictx().

Reply:
yeah, ok.  My issue was related to cleanup of dtx and then reuse.   Is this even possible in full stack?